8183569: Assert the same limits are used in parse_xss and globals.hpp

iklam opened the pull request:
We have the following in Arguments::parse_xss():

```
  const julong min_ThreadStackSize = 0;
  const julong max_ThreadStackSize = 1 * M;
  const julong min_size = min_ThreadStackSize * K;
  const julong max_size = max_ThreadStackSize * K;
```

which duplicates the min/max range specified in globals.hpp:

```
  product_pd(intx, ThreadStackSize, \
          "Thread Stack Size (in Kbytes)") \
          range(0, 1 * M) \
```

I added an API to query the min/max range of a flag, so parse_xss can be rewritten as

```
  const JVMTypedFlagLimit<intx>* limit = 
    JVMFlagLimit::get_range_at(FLAG_MEMBER_ENUM(ThreadStackSize))->cast<intx>();
  const julong min_size = limit->min() * K;
  const julong max_size = limit->max() * K;
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8183569](https://bugs.openjdk.java.net/browse/JDK-8183569): Assert the same limits are used in parse_xss and globals.hpp


### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to e28c07ccd457cb38b1bdae6f16ca0a04d493246f
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to e28c07ccd457cb38b1bdae6f16ca0a04d493246f


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2688/head:pull/2688`
`$ git checkout pull/2688`

tstuefe commented on src/hotspot/share/runtime/flags/jvmFlag.hpp:
```diff
@@ -290,6 +291,27 @@ class JVMFlag {
+    }
+#endif
+  }
+
```
Do we need this? Seems an unused duplicate to the version in jvmFlagLimit.hpp.

iklam replied:
I left `JVMFlagLimit::assert_compatible_type` by mistake, so I deleted it. I think we can use `JVMFlag::assert_compatible_type` in other places as well to get rid of the 8x duplicated boilerplate functions such as `JVMFlag::{get_bool, get_int, get_intx, etc}`. I'll investigate and file an RFE.

For your other comments, I updated `JVMFlag::assert_compatible_type` to use macros, so there's less repetition and more uniform (no more special case for `double`). It's a little easier to maintain, but maybe not as obvious to read.

tstuefe commented on src/hotspot/share/runtime/flags/jvmFlagLimit.hpp:
```diff
@@ -130,6 +137,29 @@ class JVMFlagLimit {
+      case JVMFlag::TYPE_intx:     assert(sizeof(T) == sizeof(intx)     && std::is_signed<T>::value == std::is_signed<intx>    ::value, "must be"); break;
+      case JVMFlag::TYPE_uintx:    assert(sizeof(T) == sizeof(uintx)    && std::is_signed<T>::value == std::is_signed<uintx>   ::value, "must be"); break;
+      case JVMFlag::TYPE_uint64_t: assert(sizeof(T) == sizeof(uint64_t) && std::is_signed<T>::value == std::is_signed<uint64_t>::value, "must be"); break;
+      case JVMFlag::TYPE_size_t:   assert(sizeof(T) == sizeof(size_t)   && std::is_signed<T>::value == std::is_signed<size_t>  ::value, "must be"); break;
```
Could we shorten the signed comparisons to the expected default? For most types the signedness wont be a surprise :) size_t is unsigned on all our platforms. 

is_signed<bool> seems weird, isn't that UB?

tstuefe reviewed (changes requested):
Hi Ioi,

Seems reasonable. Some remarks below.

..Thomas

iklam commented:
In rev [8c4b5cd](https://github.com/openjdk/jdk/commit/8c4b5cd9288b063482f28836ca5b0b337a412d5d):

I updated `Arguments::parse_xss()` to restore the original code -- I think it's important to see the `1 * M` to make sense of what this comment is talking about:

```
  // The min and max sizes match the values in globals.hpp, but scaled
  // with K. The values have been chosen so that alignment with page
  // size doesn't change the max value, which makes the conversions
  // back and forth between Xss value and ThreadStackSize value easier.
  // The values have also been chosen to fit inside a 32-bit signed type.
  const julong min_ThreadStackSize = 0;
  const julong max_ThreadStackSize = 1 * M;
```

So instead of the PR's initially proposed changes, I added these to make sure that the value `1 * M` matches with globals.hpp:

```
  // Make sure the above values match the range set in globals.hpp
  const JVMTypedFlagLimit<intx>* limit = JVMFlagLimit::get_range_at(FLAG_MEMBER_ENUM(ThreadStackSize))->cast<intx>();
  assert(min_ThreadStackSize == static_cast<julong>(limit->min()), "must be");
  assert(max_ThreadStackSize == static_cast<julong>(limit->max()), "must be");
```

If people are OK with this, I will update the PR's initial comment, and the bug title as well.

kimbarrett commented on src/hotspot/share/runtime/flags/jvmFlag.hpp:
```diff
@@ -290,6 +291,23 @@ class JVMFlag {
   static const char* flag_error_str(JVMFlag::Error error);
 
+  // type checking
+#define CHECK_COMPATIBLE(type) \
```
CHECK_COMPATIBLE doesn't appear to be used outside the following assertion function.  Add an `#undef` afterward.

kimbarrett commented on src/hotspot/share/runtime/flags/jvmFlag.hpp:
```diff
@@ -290,6 +291,23 @@ class JVMFlag {
+  case TYPE_##type: \
+    assert(sizeof(T) == sizeof(type) && \
+           std::is_integral<T>::value == std::is_integral<type>::value && \
+           std::is_signed  <T>::value == std::is_signed  <type>::value, "must be"); break;
```
I'd prefer the `break` be on it's own source line, rather than after the assert form.

iklam commented:
I've updated the code per @kimbarrett's suggestions, and also changed the title to "Assert the same limits are used in parse_xss and globals.hpp"

kimbarrett reviewed (approved):
Looks good.

tstuefe reviewed (approved):
Looks good to me. Sorry for the delay.

Cheers, Thomas

iklam commented:
Thanks @kimbarrett and @tstuefe for the review!
/integrate
